connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginDesc.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.runtime.isolation;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.maven.artifact.versioning.DefaultArtifactVersion;
+
+import java.util.Objects;
+
+public class PluginDesc<T> implements Comparable<PluginDesc<T>> {
+    private final Class<? extends T> klass;
+    private final String name;
+    private final String version;
+    private final DefaultArtifactVersion encodedVersion;
+    private final PluginType type;
+    private final String typeName;
+    private final String path;
+
+    public PluginDesc(Class<? extends T> klass, String version, PluginClassLoader loader) {
+        this.klass = klass;
+        this.name = klass.getCanonicalName();
+        this.version = version;
+        this.encodedVersion = new DefaultArtifactVersion(version);
+        this.type = PluginType.from(klass);
+        this.typeName = type.toString();
+        this.path = loader.path();
+    }
+
+    public Class<? extends T> pluginClass() {
+        return klass;
+    }
+
+    @Override
+    public String toString() {
+        return "PluginDesc{" +
+                "klass=" + klass +
+                ", name='" + name + '\'' +
+                ", version='" + version + '\'' +
+                ", encodedVersion=" + encodedVersion +
+                ", type=" + type +
+                ", typeName='" + typeName + '\'' +
+                ", path='" + path + '\'' +
+                '}';
+    }
+
+    @JsonProperty("class")
+    public String className() {
+        return name;
+    }
+
+    @JsonProperty("version")
+    public String version() {
+        return version;
+    }
+
+    public PluginType type() {
+        return type;
+    }
+
+    @JsonProperty("type")
+    public String typeName() {
+        return typeName;
+    }
+
+    @JsonProperty("path")
+    public String path() {
+        return typeName;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
I've been using `Objects.hash` myself. Wait I think you are contradicting yourself here. For hash I use the `Objects` version. But not for `equals`. I'd change `equals`. Anyway I was doing it a bit manually here because `compareTo` does not depend on `equals`.